DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/restricted_usm_kernel.hpp

```diff
@@ -0,0 +1,142 @@
+//  |       |    |             |  | |----------|   |               | |
+//  |-------|    |             |<---| Consumer |<==|               | |
+//               |             |  | |----------|   |---------------| |
+//               |-------------|  |                                  |
+//                                |----------------------------------|
+//
+//
+// As shown in the image above and the code below, we have split this design
+// into three kernels:
+//    1) Producer
+//    2) RestrictedUSM
+//    3) Consumer
+// We do this to decouple the reads/writes from/to the Host Memory over PCIe.
+// Decoupling the memory accesses and using SYCL pipes with a substantial
+// depth ('kPipeDepth' below) allows the kernel to be more resilient against
+// stalls while reading/writing from/to the Host Memory over PCIe. 
+//
+
+// Declare the kernel class names globally to avoid name mangling.
+class RestrictedUSM;
+class Producer;
+class Consumer;
+
+// pipes
+constexpr size_t kPipeDepth = 64;
+template <typename T>
+using ProducePipe = pipe<class ProducePipeClass, T, kPipeDepth>;
+template <typename T>
+using ConsumePipe = pipe<class ConsumePipeClass, T, kPipeDepth>;
+
+//
+// reads the input data from the hosts memory
+// and writes it to the ProducePipe
+//
+template <typename T>
+event SubmitProducer(queue& q, T* in_data, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    h.single_task<Producer>([=]() [[intel::kernel_args_restrict]] {
+      // using a host_ptr tells the compiler that this pointer lives in the
+      // hosts address space
+      host_ptr<T> h_in_data(in_data);
+
+      for (size_t i = 0; i < size; i++) {
+        const T data_from_host_memory = *(h_in_data + i);
+        ProducePipe<T>::write(data_from_host_memory);
+      }
+    });
+  });
+
+  return e;
+}
+
+//
+// The worker kernel in the device:
+//  1) read input data from the ProducePipe
+//  2) perform computation
+//  3) write the output data to the ConsumePipe
+//
+template <typename T>
+event SubmitWorker(queue& q, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    h.single_task<RestrictedUSM>([=]() [[intel::kernel_args_restrict]] {
+      for (size_t i = 0; i < size; i++) {
+        const T data = ProducePipe<T>::read();
+        const T value = data * i; // perform computation
+        ConsumePipe<T>::write(value);
+      }
+    });
+  });
+
+  return e;
+}
+
+//
+// reads output data from the device via ConsumePipe
+// and writes it to the hosts memory
+//
+template <typename T>
+event SubmitConsumer(queue& q, T* out_data, size_t size) {
+  auto e = q.submit([&](handler& h) {
+    // using a host_ptr tells the compiler that this pointer lives in the
+    // hosts address space
+    host_ptr<T> h_out_data(out_data);
+
+    h.single_task<Consumer>([=]() [[intel::kernel_args_restrict]] {
+      for (size_t i = 0; i < size; i++) {
+        const T data_to_host_memory = ConsumePipe<T>::read();
+        *(h_out_data + i) = data_to_host_memory;
```

Comment:
why not h_out_data[i] = data_to_host_memory;   ?

Reply:
host_ptr operator[] has issues.